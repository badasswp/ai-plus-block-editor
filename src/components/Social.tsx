import { __ } from '@wordpress/i18n';
import { check } from '@wordpress/icons';
import { useState, useEffect } from '@wordpress/element';
import { store as noticesStore } from '@wordpress/notices';
import { Button, TextareaControl, Icon } from '@wordpress/components';
import { useSelect, useDispatch } from '@wordpress/data';
import apiFetch from '@wordpress/api-fetch';

import Toast from '../components/Toast';

import { selectProps } from '../utils/types';
import { editorStore } from '../utils/store';

/**
 * Social.
 *
 * This Component returns the Social media
 * trending hash-tags & keywords.
 *
 * @since 1.5.0
 *
 * @return {JSX.Element} Social Hashtags.
 */
const Social = (): JSX.Element => {
	const [ social, setSocial ] = useState( '' );
	const [ isLoading, setIsLoading ] = useState( false );
	const { editPost, savePost } = useDispatch( editorStore ) as any;
	const { createErrorNotice, removeNotice } = useDispatch( noticesStore );
	const { postId, postContent, postSocial, notices } = useSelect(
		( select ) => {
			const { getNotices } = select( noticesStore );
			const {
				getCurrentPostId,
				getEditedPostContent,
				getEditedPostAttribute,
			} = select( editorStore ) as selectProps;

			return {
				postId: getCurrentPostId(),
				postContent: getEditedPostContent(),
				postSocial: getEditedPostAttribute( 'meta' )?.apbe_social,
				notices: getNotices(),
			};
		},
		[]
	);

	useEffect( () => {
		setSocial( postSocial );
	}, [ postSocial ] );

	/**
	 * This function fires when the user clicks
	 * the `Generate` button.
	 *
	 * @since 1.5.0
	 *
	 * @return { void }
	 */
	const handleClick = async (): Promise< void > => {
		notices.forEach( ( notice ) => removeNotice( notice.id ) );
		setIsLoading( true );

		try {
			const aiSocial: string = await apiFetch( {
				path: '/ai-plus-block-editor/v1/sidebar',
				method: 'POST',
				data: {
					id: postId,
					text: postContent?.text || postContent,
					feature: 'social',
				},
			} );

			/**
			 * This function returns a promise that resolves
			 * to the AI generated social media hashtags and keywords
			 * when the Animation responsible for showing same is completed.
			 *
			 * @since 1.5.0
			 *
			 * @return { Promise<string> } Animated text.
			 */
			const showAnimatedAiText = (): Promise< string > => {
				let limit = 1;

				return new Promise( ( resolve ) => {
					const animatedTextInterval = setInterval( () => {
						if ( aiSocial.length === limit ) {
							clearInterval( animatedTextInterval );
							resolve( aiSocial );
						}
						setSocial( aiSocial.substring( 0, limit ) );
						limit++;
					}, 5 );
				} );
			};

			showAnimatedAiText().then( ( newSocial ) => {
				editPost( { meta: { apbe_social: newSocial } } );
			} );

			setIsLoading( false );
		} catch ( e ) {
			createErrorNotice( e.message );
			setIsLoading( false );
		}
	};

	/**
	 * This function fires when the user selects
	 * the AI generated result.
	 *
	 * @since 1.5.0
	 *
	 * @return { void }
	 */
	const handleSelection = (): void => {
		editPost( { meta: { apbe_social: social } } );
		savePost();
	};

	return (
		<>
			<p>
				<strong>
					{ __( 'Social Media Hashtags', 'ai-plus-block-editor' ) }
				</strong>
			</p>
			<TextareaControl
				rows={ 4 }
				value={ social }
				onChange={ ( text ) => setSocial( text ) }
				__nextHasNoMarginBottom
			/>
			<div className="apbe-button-group">
				<Button variant="primary" onClick={ handleClick }>
					{ __( 'Generate', 'ai-plus-block-editor' ) }
				</Button>
				<Button variant="secondary" onClick={ handleSelection }>
					<Icon icon={ check } />
				</Button>
			</div>
			<Toast
				message={ __(
					'AI is generating text, please hold on for a bit.',
					'ai-plus-block-editor'
				) }
				isLoading={ isLoading }
			/>
		</>
	);
};

export default Social;
