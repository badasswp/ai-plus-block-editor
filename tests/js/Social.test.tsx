import '@testing-library/jest-dom';
import { act, render, fireEvent, waitFor } from '@testing-library/react';

import apiFetch from '@wordpress/api-fetch';
import { useSelect, useDispatch } from '@wordpress/data';

import Social from '../../src/components/Social';

jest.mock( '@wordpress/i18n', () => ( {
	__: jest.fn( ( arg ) => arg ),
} ) );

jest.mock( '@wordpress/data', () => ( {
	useSelect: jest.fn(),
	useDispatch: jest.fn(),
} ) );

jest.mock( '@wordpress/notices', () => ( {
	store: 'core/notices',
} ) );

jest.mock( '@wordpress/components', () => ( {
	Button: jest.fn( ( { children, variant, onClick } ) => {
		return (
			<>
				<button className={ variant } onClick={ onClick }>
					{ children }
				</button>
			</>
		);
	} ),

	Icon: jest.fn( () => {
		return <>Icon</>;
	} ),

	TextareaControl: jest.fn( ( { rows, value, onChange } ) => {
		return (
			<textarea
				rows={ rows }
				onChange={ ( e ) => onChange( e.target.value ) }
				value={ value }
			/>
		);
	} ),
} ) );

jest.mock( '@wordpress/api-fetch', () => jest.fn() );

describe( 'Social', () => {
	beforeEach( () => {
		( useSelect as jest.Mock ).mockReturnValue( {
			postId: 1,
			postContent: 'Hello World',
			postSocial: '#hello, #world',
			notices: [],
		} );

		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: jest.fn(),
			savePost: jest.fn(),
			removeNotice: jest.fn(),
			createErrorNotice: jest.fn(),
		} );

		jest.spyOn( console, 'error' ).mockImplementation( () => {} );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'renders Social component', () => {
		const { container, getByText, getByRole } = render( <Social /> );

		expect( container ).toMatchSnapshot();

		expect( getByText( '#hello, #world' ) ).toBeInTheDocument();
		expect( getByText( 'Social Media Hashtags' ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toHaveClass(
			'secondary'
		);
		expect( getByRole( 'button', { name: 'Generate' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Generate' } ) ).toHaveClass(
			'primary'
		);
	} );

	it( 'renders fetched API data from AI LLM', async () => {
		const mockEditPost = jest.fn();
		const mockCreateNotice = jest.fn();
		const mockRemoveNotice = jest.fn();

		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
			createNotice: mockCreateNotice,
			removeNotice: mockRemoveNotice,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () =>
				Promise.resolve( '#what, #beautiful, #wonderful, #world' )
			)
		);

		const { getByText, getByRole } = render( <Social /> );

		expect( getByText( '#hello, #world' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( '#what, #beautiful, #wonderful, #world' )
			).toBeInTheDocument();
		} );
	} );

	it( 'renders error notice on API fail', async () => {
		const mockCreateErrorNotice = jest.fn();
		const mockCreateNotice = jest.fn();
		const mockRemoveNotice = jest.fn();

		( useDispatch as jest.Mock ).mockReturnValue( {
			createErrorNotice: mockCreateErrorNotice,
			createNotice: mockCreateNotice,
			removeNotice: mockRemoveNotice,
		} );

		( apiFetch as unknown as jest.Mock ).mockRejectedValueOnce(
			new Error( 'AI LLM down...' )
		);

		const { getByText, getByRole } = render( <Social /> );

		expect( getByText( '#hello, #world' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockCreateErrorNotice ).toHaveBeenCalledTimes( 1 );
			expect( getByText( '#hello, #world' ) ).toBeInTheDocument();
		} );
	} );

	it( 'saves the selected AI Hashtags', async () => {
		const mockEditPost = jest.fn();
		const mockSavePost = jest.fn();
		const mockCreateNotice = jest.fn();
		const mockRemoveNotice = jest.fn();
		const mockCreateErrorNotice = jest.fn();

		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
			savePost: mockSavePost,
			createNotice: mockCreateNotice,
			removeNotice: mockRemoveNotice,
			createErrorNotice: mockCreateErrorNotice,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () =>
				Promise.resolve( '#what, #beautiful, #wonderful, #world' )
			)
		);

		const { getByText, getByRole } = render( <Social /> );

		expect( getByText( '#hello, #world' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( '#what, #beautiful, #wonderful, #world' )
			).toBeInTheDocument();
		} );

		const icon = getByRole( 'button', { name: 'Icon' } );
		await act( async () => {
			fireEvent.click( icon );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 2 );
			expect( mockSavePost ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
