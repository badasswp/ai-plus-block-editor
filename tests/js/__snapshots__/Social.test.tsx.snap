// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Social renders Social component 1`] = `
<div>
  <p>
    <strong>
      Social Media Hashtags
    </strong>
  </p>
  <textarea
    rows="4"
  >
    #hello, #world
  </textarea>
  <div
    class="apbe-button-group"
  >
    <button
      class="primary"
    >
      Generate
    </button>
    <button
      class="secondary"
    >
      Icon
    </button>
  </div>
</div>
`;
